File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/valid_parentheses/ValidParentheses.java
package main.java.com.jiang.leetcode100.valid_parentheses;

import java.util.Scanner;
import java.util.Stack;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for(int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if(ch == '(' || ch == '[' || ch == '{') {
                stack.push(ch);
            } else {
                if(stack.empty()) return false;
                char top = stack.pop();
                if(ch == ')' && top == '(' ||
                    ch == ']' && top == '[' ||
                    ch == '}' && top == '{') {
                    continue;
                }
                return false;
            }
        }

        return stack.empty();
    }
}

public class ValidParentheses {
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String s = scanner.nextLine();
        scanner.close();

        if(new Solution().isValid(s)) {
            System.out.println("True");
        } else {
            System.out.println("False");
        }
    }

}

File: Leetcode100/leetcode100_python/ValidParentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        mapping = {'(': ')', '[': ']', '{': '}'}
        stack = []
        for c in s:
            if c in mapping:
                stack.append(c)
            else:
                if not stack:
                    return False
                t = stack.pop()
                if c != mapping[t]:
                    return False
        return len(stack) == 0

if __name__ == "__main__":
    s = input()
    solution = Solution()
    print(solution.isValid(s))